### PR TITLE
Draft: Change tagged enum generation in python

### PR DESCRIPTION
**Pull request by CheatCod**

### CODE REVIEW

**Comment by MOmarMiraj** [comment]
Please make sure to verify your commits as I believe it will cause errors in the actual typeshare pipeline.

File: core/src/language/python.rs [hunk 1]
```diff
@@ -621,6 +460,155 @@ impl Python {
...
+        writeln!(
+            w,
+            "    {content_key}: Union[{}]",
+            variant_class_names.join(", ")
+        )?;
+        writeln!(w)?;
+        // write pydantic model config
+        writeln!(w, "    class Config:")?;
+        writeln!(w, "        use_enum_values = True")?;
+        Ok(())
```
Comment (MOmarMiraj) [inline core/src/language/python.rs]: would this be printed out for every single enum? I thought these are only needed for tagged enums?

File: core/src/language/python.rs [hunk 2]
```diff
@@ -754,4 +742,310 @@ mod test {
...
+            Unit,
+            Tuple(i32),
+            Struct {
+                #[serde(default)]
+                field: String,
+            },
+        }
+
+        let shared_enum = RustEnumShared {
+            id: Id {
```
Comment (MOmarMiraj) [inline core/src/language/python.rs]: Is there anyway we can use the `parse_enum` function inside the `parser.rs`? I am scared that directly instantiating the `RustEnumShared` is prone to human errors.

File: core/src/language/python.rs (hunk 1 above)
Comment (CheatCod) [reply core/src/language/python.rs]: This would be printed out for every algebraic enum, which I believe should all be tagged.

**Comment by MOmarMiraj** [comment]
Just ran the test and this is what it prints:
```
class TestStructInner(BaseModel):
    """
    Generated type representing the anonymous struct variant `Struct` of the `Test` Rust enum
    """
    field: str


class TestTypes(str, Enum):
        UNIT = "Unit",
        TUPLE = "Tuple",
        STRUCT = "Struct"

TestUnit(BaseModel):
    content = Literal["Unit"]

class Tuple(BaseModel):
    content: int

class Test(BaseModel):
    type: TestTypes
    content: Union[TestUnit, TestTuple, TestStruct]

    class Config:
        use_enum_values = True
```
 Couple of issues so far:
        1) `TestUnit` doesn't have class beside it 
        2) Inside `Test`, content defines a Union with three types but only `TestUnit` is defined
        3) We have a `TestStructInner` class but no helper functions to return a `TestStructInner`
        4) Dont think we need the commas in class `TestTypes`
        
  I think it should be something like this:

```from pydantic import BaseModel
from enum import Enum
from typing import Union, Literal

class TestStructInner(BaseModel):
    """
    Generated type representing the anonymous struct variant `Struct` of the `Test` Rust enum
    """
    field: str

class TestTypes(str, Enum):
    UNIT = "Unit"
    TUPLE = "Tuple"
    STRUCT = "Struct"

class TestUnit(BaseModel):
    content: Literal["Unit"]

class TestTuple(BaseModel):
    content: int

class TestStruct(BaseModel):
    content: TestStructInner

class Test(BaseModel):
    type: TestTypes
    content: Union[TestUnit, TestTuple, TestStruct]

    class Config:
        use_enum_values = True

    @staticmethod
    def new_struct_variant(content: str) -> "Test":
        """
        Static method to create a `Test` instance with the `STRUCT` variant and populate `TestStructInner` field.
        """
        return Test(
            type=TestTypes.STRUCT,
            content=TestStruct(content=TestStructInner(field=content))
        )
```
or something similar WDYT?

**Comment by MOmarMiraj** [comment]
Currently the tests just ensure that the enum is correctly getting translated which is not exactly what we want from this tests. We want to ensure that the generation of our algebraic enums are correct with the correct imports etc.

My suggestion is to create an `expected_algebraic_enum.py` file and compare it to the one generated in this test. There are examples from the `typeshare` repo that do this inside of `core/data/tests`. We should follow that similar format for these type of tests. That way we can see if the imports are being generated correctly as right now we have to manually check that. I don't think unit tests will benefit as much as our full e2e tests will like in `core/data/tests`

**Comment by CheatCod** [comment]
With the current code, the `ItemModification` struct in Rust:

```rs
pub enum ItemModification {
    SetTitle {
        new_title: String,
    },
    AddField {
        field_label: String,
        field_value: String,
        field_type: String,
        section_title: Option<String>,
    },
    SetField {
        field_name: String,
        new_value: String,
        section_title: Option<String>,
    },
    RemoveField {
        field_name: String,
        section_title: Option<String>,
    },
}

```

produces the following python code:

```py
class ItemModificationSetTitleInner(BaseModel):
    """
    Generated type representing the anonymous struct variant `SetTitle` of the `ItemModification` Rust enum
    """
    new_title: str


class ItemModificationAddFieldInner(BaseModel):
    """
    Generated type representing the anonymous struct variant `AddField` of the `ItemModification` Rust enum
    """
    field_label: str
    field_value: str
    field_type: str
    section_title: Optional[str]


class ItemModificationSetFieldInner(BaseModel):
    """
    Generated type representing the anonymous struct variant `SetField` of the `ItemModification` Rust enum
    """
    field_name: str
    new_value: str
    section_title: Optional[str]


class ItemModificationRemoveFieldInner(BaseModel):
    """
    Generated type representing the anonymous struct variant `RemoveField` of the `ItemModification` Rust enum
    """
    field_name: str
    section_title: Optional[str]


class ItemModificationTypes(str, Enum):
    SET_TITLE = "SetTitle"
    ADD_FIELD = "AddField"
    SET_FIELD = "SetField"
    REMOVE_FIELD = "RemoveField"

class ItemModification(BaseModel):
    type: ItemModificationTypes
    content: Union[ItemModificationSetTitleInner, ItemModificationAddFieldInner, ItemModificationSetFieldInner, ItemModificationRemoveFieldInner]

    class Config:
        use_enum_values = True
```

Which can be constructed like so:

```py
item_modification = ItemModification(
	type=ItemModificationTypes.SET_TITLE,
	content=ItemModificationSetTitleInner(new_title="New Title")
)
```

**Comment by MOmarMiraj** [comment]
> With the current code, the `ItemModification` struct in Rust:
> 
> ```rust
> pub enum ItemModification {
>     SetTitle {
>         new_title: String,
>     },
>     AddField {
>         field_label: String,
>         field_value: String,
>         field_type: String,
>         section_title: Option<String>,
>     },
>     SetField {
>         field_name: String,
>         new_value: String,
>         section_title: Option<String>,
>     },
>     RemoveField {
>         field_name: String,
>         section_title: Option<String>,
>     },
> }
> ```
> 
> produces the following python code:
> 
> ```python
> class ItemModificationSetTitleInner(BaseModel):
>     """
>     Generated type representing the anonymous struct variant `SetTitle` of the `ItemModification` Rust enum
>     """
>     new_title: str
> 
> 
> class ItemModificationAddFieldInner(BaseModel):
>     """
>     Generated type representing the anonymous struct variant `AddField` of the `ItemModification` Rust enum
>     """
>     field_label: str
>     field_value: str
>     field_type: str
>     section_title: Optional[str]
> 
> 
> class ItemModificationSetFieldInner(BaseModel):
>     """
>     Generated type representing the anonymous struct variant `SetField` of the `ItemModification` Rust enum
>     """
>     field_name: str
>     new_value: str
>     section_title: Optional[str]
> 
> 
> class ItemModificationRemoveFieldInner(BaseModel):
>     """
>     Generated type representing the anonymous struct variant `RemoveField` of the `ItemModification` Rust enum
>     """
>     field_name: str
>     section_title: Optional[str]
> 
> 
> class ItemModificationTypes(str, Enum):
>     SET_TITLE = "SetTitle"
>     ADD_FIELD = "AddField"
>     SET_FIELD = "SetField"
>     REMOVE_FIELD = "RemoveField"
> 
> class ItemModification(BaseModel):
>     type: ItemModificationTypes
>     content: Union[ItemModificationSetTitleInner, ItemModificationAddFieldInner, ItemModificationSetFieldInner, ItemModificationRemoveFieldInner]
> 
>     class Config:
>         use_enum_values = True
> ```
> 
> Which can be constructed like so:
> 
> ```python
> item_modification = ItemModification(
> 	type=ItemModificationTypes.SET_TITLE,
> 	content=ItemModificationSetTitleInner(new_title="New Title")
> )
> ```

The UX looks really nice, sorry I feel like the addition of `Inner` isn't nice, we should probably have it like 
```
item_modification = ItemModification(
	type=ItemModificationTypes.SET_TITLE,
	content=ItemModificationSetTitle(new_title="New Title")
)
```
This would be ideal.


Another idea to make it nicer UX would be helper functions on `ItemModification` so the user would just have to put the argument. Like this:
```
    @classmethod
    def set_title(cls, new_title: str) -> ItemModification:
        return cls(
            type=ItemModificationTypes.SET_TITLE,
            content=ItemModificationSetTitleInner(new_title=new_title)
        )
```
this way a dev can call this:
```
    item_modifications = ItemModification[ItemModification.set_title(new_title="jep"),ItemModification.remove_field(field_title="test")]
```

**Comment by MOmarMiraj** [comment]
Also, I did some additional testing, it looks like if there is tagged enums, the correct imports are not added (Union and Enum). 

We should also have a test where the `input.rs` is just an enum and see if the other imports are added like `Optional` and `BaseModel`

File: core/data/tests/anonymous_struct_with_rename/output.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+"""
+ Generated by typeshare 1.11.0
+"""
+from __future__ import annotations
+
+from pydantic import BaseModel, ConfigDict, Enum, Field
```
Comment (MOmarMiraj) [inline core/data/tests/anonymous_struct_with_rename/output.py]: This is for all snapshots but pydantic doesn't have an `Enum` import. I believe it has to be:

`from enum import Enum`

and this goes for all snapshots.

File: core/data/tests/can_apply_prefix_correctly/output.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+"""
+ Generated by typeshare 1.11.0
+"""
+from __future__ import annotations
+
+from pydantic import BaseModel, Enum
```
Comment (MOmarMiraj) [inline core/data/tests/can_apply_prefix_correctly/output.py]: When theres algebraic enums, we should also include the `ConfigDict` import from `pydantic`

File: core/data/tests/can_generate_generic_enum/output.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    """
+    non_generic: int
+
+
+class GenericEnumsUsingStructVariantsVariantI(GenericModel, Generic[T, U]):
+    """
+    Generated type representing the anonymous struct variant `VariantI` of the `GenericEnumsUsingStructVariants` Rust enum
+    """
+    vec: List[T]
+    action: MyType[T, U]
```
Comment (MOmarMiraj) [inline core/data/tests/can_generate_generic_enum/output.py]: For all algebraic enums (beside unit) I think we should add those helper functions I talked about before.

File: core/data/tests/can_generate_anonymous_struct_with_skipped_fields/output.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class AutofilledByUs(BaseModel):
+    """
+    Generated type representing the anonymous struct variant `Us` of the `AutofilledBy` Rust enum
+    """
+    uuid: str
+    """
+    The UUID for the fill
+    """
```
Comment (MOmarMiraj) [inline core/data/tests/can_generate_anonymous_struct_with_skipped_fields/output.py:18]: I don't know if we test but according to the python styling guide, PEP8 we should uppercase acronyms if they are apart of the class and lowercase if they are a field. Maybe we want to add python to the uppercase acronym test?

Maybe we do account for it but a test would be nice to increase confidence.

File: core/data/tests/anonymous_struct_with_rename/output.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+"""
```
Comment (MOmarMiraj) [inline core/data/tests/anonymous_struct_with_rename/output.py:1]: Not related to this test, but theres still a lot of tests from the original typeshare repo, that the Python TS doesn't test.

If we don't directly support it and its a lot of work, we can note them and then when we share to external stakeholders we can say that x functionality is missing but is not necessary and issue x is created for it. As a lot of these tests are super duper edge cases.

Comment (MOmarMiraj) [reply core/data/tests/anonymous_struct_with_rename/output.py:1]: spoke with @CheatCod don't want to introduce too much scope creep, these addtiional tests can be tackled with a later PR. This doesn't reduce confidence as those are super nit tests that wont affect the average typeshare user.

**Comment by hculea** [comment]
```
item_modification = ItemModification(
	type=ItemModificationTypes.SET_TITLE,
	content=ItemModificationSetTitleInner(new_title="New Title")
)
```

I would not like that we expose this complexity to the user. If we do, this is letting an internal, technical detail (the shape of the structs that we use for serialization/deserialization) dictate the DevEx, which is a pattern we should move away from. 

Instead, I suggest we incorporate in our typeshare implementation constructors too, similar to how Go typeshare does it. For each variant, also generate a function such as `SetTitleItemModification(title)` that converts to this ugly type: https://github.com/1Password/typeshare/blob/main/core/src/language/go.rs#L358

**Comment by hculea** [comment]
Question: what happens when some variants have struct content, some have anonymous content and some don't?

For example:
```
pub enum MyCustomEnum {
    VariantOne,
    VariantTwo(ItemOverview)
    VariantThree {
          item: Item
    }
}
```
?

Can you please post the output of running this, and show what the UX would be for the user constructing each of these?

Go nicely handles the first scenario with parameter-less constructors:
```
func NewMyCustomEnumTypeVariantOne() MyCustomEnum {
	return MyCustomEnum{
		Type:    MyCustomEnumVariantOne,
		content: nil,
	}
}
```

**Comment by CheatCod** [comment]
For 
```rs
/// This is a comment (yareek sameek wuz here)
#[typeshare]
#[serde(tag = "type", content = "content")]
pub enum EnumWithManyVariants {
    UnitVariant,
    TupleVariantString(String),
    AnonVariant { uuid: String },
    TupleVariantInt(i32),
    AnotherUnitVariant,
    AnotherAnonVariant { uuid: String, thing: i32 },
}
```

It generates 
```py
class EnumWithManyVariantsAnonVariant(BaseModel):
    """
    Generated type representing the anonymous struct variant `AnonVariant` of the `EnumWithManyVariants` Rust enum
    """
    uuid: str


class EnumWithManyVariantsAnotherAnonVariant(BaseModel):
    """
    Generated type representing the anonymous struct variant `AnotherAnonVariant` of the `EnumWithManyVariants` Rust enum
    """
    uuid: str
    thing: int


class EnumWithManyVariantsTypes(str, Enum):
    UNIT_VARIANT = "UnitVariant"
    TUPLE_VARIANT_STRING = "TupleVariantString"
    ANON_VARIANT = "AnonVariant"
    TUPLE_VARIANT_INT = "TupleVariantInt"
    ANOTHER_UNIT_VARIANT = "AnotherUnitVariant"
    ANOTHER_ANON_VARIANT = "AnotherAnonVariant"

class EnumWithManyVariantsUnitVariant(BaseModel):
    content = Literal["UnitVariant"]

class EnumWithManyVariantsTupleVariantString(BaseModel):
    content: str

class EnumWithManyVariantsTupleVariantInt(BaseModel):
    content: int

class EnumWithManyVariantsAnotherUnitVariant(BaseModel):
    content = Literal["AnotherUnitVariant"]

class EnumWithManyVariants(BaseModel):
    model_config = ConfigDict(use_enum_values=True)
    type: EnumWithManyVariantsTypes
    content: Union[EnumWithManyVariantsUnitVariant, EnumWithManyVariantsTupleVariantString, EnumWithManyVariantsAnonVariant, EnumWithManyVariantsTupleVariantInt, EnumWithManyVariantsAnotherUnitVariant, EnumWithManyVariantsAnotherAnonVariant]


    @classmethod
    def AnonVariant(cls, uuid: str):
        return cls(
		type=EnumWithManyVariantsTypes.ANON_VARIANT,
		content=EnumWithManyVariantsAnonVariant(uuid = uuid))



    @classmethod
    def AnotherAnonVariant(cls, uuid: str, thing: int):
        return cls(
		type=EnumWithManyVariantsTypes.ANOTHER_ANON_VARIANT,
		content=EnumWithManyVariantsAnotherAnonVariant(uuid = uuid, thing = thing))

```

No ctor on non-struct variant yet, i will work on that

**Comment by MOmarMiraj** [comment]
Paired with @CheatCod to add majority of the tests and Peter fixed up the remaining work needed for the helper functions in tagged enums. 

@hculea this should be ready for a full review but outputs were looked over by myself and Peter and they looked like Idiomatic Python.

Some snapshot tests that we don't support currently is:
```
can_override_types - we don't support that macro yet
all the generic tests - We don't support generics in Python TS yet due to needing more research on whats idiomatic and isn't super important.
serde_default - we don't support adding a default in Python 
```

File: core/src/language/python.rs [hunk 8]
```diff
@@ -275,27 +275,31 @@ impl Language for Python {
             // Write all the unit variants out (there can only be unit variants in
             // this case)
             RustEnum::Unit(shared) => {
-                self.add_import("typing".to_string(), "Literal".to_string());
-                write!(
-                    w,
-                    "{} = Literal[{}]",
-                    shared.id.renamed,
+                self.add_import("enum".to_string(), "Enum".to_string());
```
Comment (MOmarMiraj) [inline core/src/language/python.rs]: We can just have this at the start of the function because no matter what we need to import this. For unit or algebraic enums.

File: core/src/language/python.rs [hunk 9]
```diff
@@ -621,6 +458,318 @@ impl Python {
...
+        make_struct_name: &dyn Fn(&str) -> String,
+    ) -> std::io::Result<()> {
+        shared
+            .generic_types
+            .iter()
+            .cloned()
+            .for_each(|v| self.add_type_var(v));
+        let mut variants: Vec<(String, Vec<String>)> = Vec::new();
+        self.add_import("typing".to_string(), "Union".to_string());
+        self.add_import("enum".to_string(), "Enum".to_string());
```
Comment (MOmarMiraj) [inline core/src/language/python.rs]: We can remove it here!

File: core/src/language/python.rs [hunk 10]
```diff
@@ -579,20 +430,6 @@ impl Python {
...
-                    "T".to_string()
-                } else {
-                    format!("T{}", i)
-                }
-            })
-            .collect();
-        vars.iter().for_each(|tv| self.add_type_var(tv.clone()));
-        vars
-    }
-
```
Comment (MOmarMiraj) [inline core/src/language/python.rs:595]: this wasn't used so I deleted it.

**Comment by MOmarMiraj** [comment]
Just looking the fact we don't support serde default brings up concerns about how we are going about it.

The way typeshare defines the `is_default` field is as follows:
```
    /// This will be true if the field has a `serde(default)` decorator.
    /// Even if the field's type is not optional, we need to make it optional
    /// for the languages we generate code for.
    pub has_default: bool,
```
The way we currently have it is that we only set a field to `Optional<T> = None` if the field is Optional and has the `#[serde(default)]` tag on it. This goes against what typeshare has set out for the `has_default` field which we should adopt.

I believe what we are trying to do currently is provide the default value for the field that has the `#[serde(default)]` tag on it (i.e. if its a vector, the default value is an empty vector). To my understanding, if we omit the field and deserialize the struct, serde fills in the missing field with the default value, which is why typeshare has opted to make fields with the  `#[serde(default)]`  tag act like an option.

WDYT @hculea @CheatCod

File: core/data/tests/can_generate_anonymous_struct_with_skipped_fields/output.py (hunk 6 above)
Comment (MOmarMiraj) [reply core/data/tests/can_generate_anonymous_struct_with_skipped_fields/output.py:18]: We can leave this as an improvement to be made in the external PR.

File: core/src/language/python.rs [hunk 11]
```diff
@@ -621,6 +458,318 @@ impl Python {
         };
         Ok(())
     }
+
+    fn gen_unit_variant_ctor(
```
Comment (MOmarMiraj) [inline core/src/language/python.rs]: instead of ctor can we put constructor just so we dont get confused and that goes for all functions.

Comment (MOmarMiraj) [inline core/src/language/python.rs]: Also for the three constructor functions, theres some duplication that we can clean up like the method name and even the writing of the functions

File: core/src/language/python.rs [hunk 12]
```diff
@@ -621,6 +458,318 @@ impl Python {
...
+                        RustType::Generic { parameters, .. } => {
+                            let mut generic_parameters: Vec<String> = parameters
+                                .iter()
+                                .flat_map(|p| {
+                                    collect_generics_for_variant(p, &shared.generic_types)
+                                })
+                                .collect();
+                            dedup(&mut generic_parameters);
+                            if generic_parameters.is_empty() {
+                                self.add_import("pydantic".to_string(), "BaseModel".to_string());
```
Comment (MOmarMiraj) [inline core/src/language/python.rs]: I think base model should imported for any type we pass through so I think we can push that to a higher level function instead of here.

**Review by MOmarMiraj — COMMENTED** [review]
Left some more comments, I believe if we tackle those and get confirmation from Horia about the outputs and `serde(default)` than we should be in a position to merge and ask for review today!

Nice Job!

**Comment by hculea** [comment]
For this:
```
#[typeshare]
#[serde(tag = "type", content = "content", rename_all = "camelCase")]
pub enum MoreOptions {
    Hello,
    News(bool),
    Exactly { config: String },
    Built { top: MoreOptions },
}
```

we also generate:
```
class MoreOptionsHello(BaseModel):
    content = Literal["hello"]
```
and then use it in the union:
```
Union[MoreOptionsHello, MoreOptionsNews, MoreOptionsExactly, MoreOptionsBuilt, None]
```

but this is actually redundant (and it triggers a pydantic error) - instead, we should not generate this at all, and simply use `None` as the content type in that case.

**Comment by hculea** [comment]
To @MOmarMiraj 's point above - yes, let's make sure that we respect the typeshare spec (and make the optional fields optional despite only having `serde(default)`. I'll let you two assess whether this should be in the scope of this PR. We could do it if it's small enough, but this is already quite sizeable as is, so I'd suggest we avoid additional scope creep.

**Comment by MOmarMiraj** [comment]
Everything looks good! Approving for functional test and code!
